File: vscode-extension/scripts/src/modules/converter.py
"""
转换器主模块
提供从Markdown到Word的完整转换功能
"""

import os
import codecs
from typing import Dict, Any, Optional, List, Union
from docx import Document

# 使用try-except处理不同的导入场景
try:
    # 相对导入（作为包的一部分被导入时）
    from .markdown_to_html import MarkdownToHtml
    from .html_to_word import HtmlToWordConverter
    from .html_elements_processor import HtmlElementsProcessor
except ImportError:
    try:
        # 绝对导入
        from src.modules.markdown_to_html import MarkdownToHtml
        from src.modules.html_to_word import HtmlToWordConverter
        from src.modules.html_elements_processor import HtmlElementsProcessor
    except ImportError:
        # 从当前目录导入
        from markdown_to_html import MarkdownToHtml
        from html_to_word import HtmlToWordConverter
        from html_elements_processor import HtmlElementsProcessor

class Converter:
    """
    /**
     * 转换器主类
     * 
     * 整合Markdown到HTML和HTML到Word的转换功能，
     * 提供完整的文档转换流程
     */
    """
    
    def __init__(self, config: Dict[str, Any]):
        """
        /**
         * 初始化转换器
         * 
         * @param {Dict[str, Any]} config - 配置参数字典
         */
        """
        self.config = config
        self.md_to_html = MarkdownToHtml(config)
        self.html_to_word = HtmlToWordConverter(config)
        self.html_processor = HtmlElementsProcessor(config)
        
    def convert_file(self, input_file: str, output_file: str, keep_html: bool = False) -> Document:
        """
        /**
         * 转换单个Markdown文件为Word文档
         * 
         * @param {str} input_file - 输入Markdown文件路径
         * @param {str} output_file - 输出Word文件路径
         * @param {bool} keep_html - 是否保留中间HTML文件
         * @returns {Document} 生成的Word文档对象
         */
        """
        if not os.path.exists(input_file):
            raise FileNotFoundError(f"输入文件不存在: {input_file}")
            
        # 确定HTML中间文件路径
        html_file = None
        if keep_html:
            base_name = os.path.basename(os.path.splitext(input_file)[0])
            # 指向根目录的html文件夹
            html_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "html")
            os.makedirs(html_dir, exist_ok=True)
            html_file = os.path.join(html_dir, f"{base_name}.html")
            
        # 转换Markdown到HTML
        html_content = self.md_to_html.convert_file(input_file, html_file)
        
        # 转换HTML到Word
        if html_file:
            doc = self.html_to_word.convert_file(html_file, output_file)
        else:
            doc = self.html_to_word.convert_html(html_content)
            doc.save(output_file)
            
        return doc
        
    def convert_text(self, md_content: str, output_file: Optional[str] = None) -> Union[str, Document]:
        """
        /**
         * 转换Markdown文本内容
         * 
         * @param {str} md_content - Markdown格式的文本
         * @param {Optional[str]} output_file - 输出Word文件路径，如果不提供则不保存文件
         * @returns {Union[str, Document]} 如果提供output_file则返回Document对象，否则返回HTML内容
         */
        """
        # 转换Markdown到HTML
        html_content = self.md_to_html.convert_text(md_content)
        
        # 如果没有指定输出文件，直接返回HTML内容
        if not output_file:
            return html_content
            
        # 转换HTML到Word并保存
        doc = self.html_to_word.convert_html(html_content)
        doc.save(output_file)
        
        return doc
        
    def batch_convert(self, input_dir: str, output_dir: str, keep_html: bool = False) -> Dict[str, bool]:
        """
        /**
         * 批量转换目录中的Markdown文件
         * 
         * @param {str} input_dir - 输入目录路径
         * @param {str} output_dir - 输出目录路径
         * @param {bool} keep_html - 是否保留中间HTML文件
         * @returns {Dict[str, bool]} 文件转换结果字典，键为文件名，值为转换是否成功
         */
        """
        if not os.path.exists(input_dir):
            raise FileNotFoundError(f"输入目录不存在: {input_dir}")
            
        # 确保输出目录存在
        os.makedirs(output_dir, exist_ok=True)
        
        # 如果保留HTML文件，创建/使用根目录下的HTML目录
        html_dir = None
        if keep_html:
            # 指向根目录的html文件夹
            html_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "html")
            os.makedirs(html_dir, exist_ok=True)
            
        # 查找所有Markdown文件
        results = {}
        files = self._find_markdown_files(input_dir)
        
        # 转换每个文件
        total_files = len(files)
        for idx, file_path in enumerate(files, 1):
            # 计算相对路径，用于构建输出路径
            rel_path = os.path.relpath(file_path, input_dir)
            file_base_name = os.path.basename(os.path.splitext(rel_path)[0])
            base_name = os.path.splitext(rel_path)[0]
            
            # 构建输出文件路径
            output_file = os.path.join(output_dir, f"{base_name}.docx")
            
            # 确保输出目录存在
            os.makedirs(os.path.dirname(output_file), exist_ok=True)
            
            # 构建HTML文件路径（如果需要）
            html_file = None
            if keep_html:
                html_file = os.path.join(html_dir, f"{file_base_name}.html")
                
            # 转换文件并记录结果
            try:
                # 输出进度信息
                print(f"处理文件 {idx}/{total_files}: {rel_path}")
                
                # 转换Markdown到HTML
                html_content = self.md_to_html.convert_file(file_path, html_file)
                
                # 转换HTML到Word
                if html_file:
                    doc = self.html_to_word.convert_file(html_file, output_file)
                else:
                    doc = self.html_to_word.convert_html(html_content)
                    doc.save(output_file)
                    
                results[rel_path] = True
                print(f"  完成: {output_file}")
                
            except Exception as e:
                results[rel_path] = False
                print(f"  失败: {str(e)}")
                
        # 输出统计信息
        success_count = sum(1 for v in results.values() if v)
        print(f"\n转换完成: 共 {total_files} 个文件, 成功 {success_count} 个, 失败 {total_files - success_count} 个")
        
        return results
    
    def _find_markdown_files(self, directory: str) -> List[str]:
        """
        /**
         * 在目录中查找所有Markdown文件
         * 
         * @param {str} directory - 要搜索的目录
         * @returns {List[str]} Markdown文件路径列表
         */
        """
        md_files = []
        
        for root, _, files in os.walk(directory):
            for file in files:
                if file.lower().endswith(('.md', '.markdown')):
                    md_files.append(os.path.join(root, file))
                    
        return md_files
    
    def cleanup(self):
        """
        /**
         * 清理临时资源
         */
        """
        # 清理HTML处理器的临时资源
        self.html_processor.cleanup() 

File: vscode-extension/src/mdtoworld/converter.js
/**
 * @description 主转换器模块
 * 整合Markdown到HTML和HTML到Word的转换功能
 */
const fs = require('fs-extra');
const path = require('path');
const glob = require('glob');
const { Packer } = require('docx');

const MarkdownToHtml = require('./markdownToHtml');
const { HtmlToWordConverter } = require('./htmlToWord');
const ConfigManager = require('./utils/configManager');
const FileHandler = require('./utils/fileHandler');
const ErrorHandler = require('./utils/errorHandler');
const { COMPLETE_DEFAULTS, deepMerge } = require('./utils/defaultConfig');

/**
 * @class Converter
 * @description 主转换器类，协调Markdown到HTML，HTML到Word的转换过程
 */
class Converter {
  /**
   * @constructor
   * @param {Object|string} config - 配置对象或配置文件路径
   */
  constructor(config = {}) {
    // console.log('初始化转换器...');
    
    // 初始化配置（同步方式）
    this.configManager = new ConfigManager();
    
    if (typeof config === 'string') {
      // 如果是字符串，设置为配置文件路径，延迟加载
      this.configPath = config;
      this.config = { ...COMPLETE_DEFAULTS }; // 临时默认配置
    } else {
      // 深度合并用户配置与完整默认值，确保所有键存在
      this.config = deepMerge(COMPLETE_DEFAULTS, config || {});
      this.configManager = new ConfigManager(this.config);
    }
    
    // 初始化文件处理器
    this.fileHandler = new FileHandler({
      maxRetries: 3,
      retryDelay: 1000,
      autoRename: true
    });
    
    // 创建转换器实例
    this.md_to_html = new MarkdownToHtml(this.config);
    this.html_to_word = new HtmlToWordConverter(this.config);
  }
  
  /**
   * @method initializeConfigAsync
   * @description 异步初始化配置（如果需要从文件加载）
   * @private
   */
  async initializeConfigAsync() {
    if (this.configPath) {
      try {
        const loaded = await this.configManager.loadFromYaml(this.configPath);
        if (!loaded) {
          // console.warn(`无法加载配置文件: ${this.configPath}，将使用默认配置`);
        } else {
          this.config = deepMerge(COMPLETE_DEFAULTS, this.configManager.getAll());
          // 更新转换器的配置
          this.md_to_html = new MarkdownToHtml(this.config);
          this.html_to_word = new HtmlToWordConverter(this.config);
        }
      } catch (error) {
        // console.error(`加载配置文件失败: ${error.message}`);
      }
    }
  }

  /**
   * @method convert_file
   * @description 转换单个Markdown文件为Word文档
   * @param {string} input_file - 输入Markdown文件路径
   * @param {string} output_file - 输出Word文件路径
   * @param {boolean} keep_html - 是否同时产出 gov 样式 HTML 文档（默认 true）
   * 说明：HTML 与 Word 一同产出到输出目录（同名 .html），用户可按需选择 HTML 或 Word
   * @returns {Object} - 转换结果
   */
  async convert_file(input_file, output_file, keep_html = true) {
    try {
      // 首先初始化配置（如果有配置文件路径）
      await this.initializeConfigAsync();
      
      if (!await fs.pathExists(input_file)) {
        throw new Error(`输入文件不存在: ${input_file}`);
      }

      // console.log(`开始转换文件: ${input_file} -> ${output_file}`);

      // 确定 HTML 输出路径：与 Word 输出同目录、同名（.md → .html），而非源文件目录
      let html_file = null;
      if (keep_html) {
        // 以 Word 输出文件名为基准（同名配套），保证用户拿到的 .html 与 .docx 一一对应
        const out_base = path.basename(output_file, path.extname(output_file));
        html_file = path.join(path.dirname(output_file), `${out_base}.html`);
      }

      // 确保输出目录存在（HTML 与 Word 同目录）
      await fs.ensureDir(path.dirname(output_file));

      // 打印配置信息以便调试
      // console.log('使用配置转换Markdown:', JSON.stringify({
      //   chinese: this.config.chinese,
      //   document: this.config.document,
      //   paragraph: this.config.paragraph,
      //   table_styles: this.config.table_styles,
      //   enhanced_table_styles: this.config.enhanced_table_styles
      // }, null, 2));

      // 确保配置中的中文相关设置传递给markdown到html模块
      // 每次创建新实例避免配置共享问题
      this.md_to_html = new MarkdownToHtml(this.config);

      // ===== 转换 Markdown → HTML（Node 服务端渲染完成） =====
      // 用户可见 .html 与 Word 内部均用 convertString：服务端 marked 已渲染静态 DOM
      // （gov 公文样式 + mermaid SVG + 图片放大增强），打开即正确文档，不依赖浏览器 JS
      // 注意：不要用 convertGovFile（浏览器 JS 渲染 md-source + marked.parse）——
      // 否则 HTML 打开时禁用/未执行 JS 会显示 markdown 原文或空白
      let html_content;
      try {
        // 只转换一次：服务端渲染完成（含 mermaid 离线渲染），结果同时用于
        // ① 用户可见 .html 文件 ② Word 内部 htmlToWord 解析
        // （避免对同一 md 做两次转换导致 mermaid 渲染两遍、耗时翻倍）
        html_content = await this.md_to_html.convertFile(input_file);
        if (html_file) {
          await fs.outputFile(html_file, html_content, 'utf-8');
        }
      } catch (error) {
        throw ErrorHandler.createUserFriendlyError(error, 'markdown-to-html');
      }

      // 转换HTML到Word
      let doc;
      let writeResult;
      
      // Word 转换统一用内存中的静态 HTML（convertHtml 解析 DOM），
      // 不再读 html_file（那是给用户看的 gov 浏览器渲染版，含 <script> 无静态 DOM）
      try {
        // console.log('开始HTML到Word转换...');
        doc = await this.html_to_word.convertHtml(html_content, path.dirname(input_file));
        
        // 创建输出目录（如果不存在）
        await fs.ensureDir(path.dirname(output_file));
        
        // 使用文件处理器保存Word文档
        // console.log('生成Word文档缓冲区...');
        const buffer = await Packer.toBuffer(doc);
        
        // console.log(`准备保存Word文档到: ${output_file}`);
        writeResult = await this.fileHandler.writeWordDocument(output_file, buffer);
        
        if (!writeResult.success) {
          const error = writeResult.error || new Error(writeResult.message);
          error.code = writeResult.error?.code;
          throw ErrorHandler.createUserFriendlyError(error, 'file-save', output_file);
        }
        
        // 更新实际的输出文件路径（可能因为重命名而改变）
        output_file = writeResult.filePath;
      } catch (error) {
        if (error.errorInfo) {
          // 已经是用户友好的错误，直接抛出
          throw error;
        } else if (error.code === 'EBUSY' || error.code === 'EACCES' || error.code === 'EPERM') {
          throw ErrorHandler.createUserFriendlyError(error, 'file-save', output_file);
        } else {
          throw ErrorHandler.createUserFriendlyError(error, 'html-to-word');
        }
      }

      const successMessage = writeResult ? writeResult.message : '转换成功';
      // console.log(`✅ ${successMessage}`);

      return {
        success: true,
        message: successMessage,
        outputFile: output_file,
        htmlFile: html_file,
        document: doc
      };
    } catch (error) {
      // console.error('转换文件时出错:', error.originalError || error);
      
      return {
        success: false,
        message: error.message,
        error: error.originalError || error
      };
    }
  }

  /**
   * @method batch_convert
   * @description 批量转换目录中的Markdown文件为Word文档
   * @param {string} input_dir - 输入目录路径
   * @param {string} output_dir - 输出目录路径
   * @param {boolean} keep_html - 是否保留中间HTML文件
   * @returns {Object} - 转换结果，键为文件路径，值为是否成功
   */
  async batch_convert(input_dir, output_dir, keep_html = false) {
    try {
      if (!await fs.pathExists(input_dir)) {
        throw new Error(`输入目录不存在: ${input_dir}`);
      }

      // 确保输出目录存在
      await fs.ensureDir(output_dir);

      // 查找所有Markdown文件
      const files = glob.sync(path.join(input_dir, '**', '*.md'));
      const results = {};

      // 转换每个文件
      const total_files = files.length;
      for (let idx = 0; idx < total_files; idx++) {
        const file_path = files[idx];
        
        // 计算相对路径，用于构建输出路径
        const rel_path = path.relative(input_dir, file_path);
        const base_name = path.basename(rel_path, '.md');
        const dir_name = path.dirname(rel_path);
        
        // 构建输出文件路径
        const output_file = path.join(output_dir, dir_name, `${base_name}.docx`);
        
        // 确保输出目录存在
        await fs.ensureDir(path.dirname(output_file));
        
        // 输出进度信息
        // console.log(`处理文件 ${idx + 1}/${total_files}: ${rel_path}`);
        
        try {
          // 执行转换
          const result = await this.convert_file(file_path, output_file, keep_html);
          results[rel_path] = result.success;
          // console.log(`  ${result.success ? '完成' : '失败'}: ${output_file}`);
        } catch (error) {
          results[rel_path] = false;
          // console.error(`  失败: ${error.message}`);
        }
      }

      // 输出统计信息
      const success_count = Object.values(results).filter(v => v).length;
      // console.log(`\n转换完成: 共 ${total_files} 个文件, 成功 ${success_count} 个, 失败 ${total_files - success_count} 个`);
      
      return results;
    } catch (error) {
      // console.error('批量转换时出错:', error);
      throw error;
    }
  }

  /**
   * @method loadConfig
   * @description 加载配置文件
   * @param {string} configPath - 配置文件路径
   * @returns {boolean} - 是否成功加载
   */
  async loadConfig(configPath) {
    try {
      const loaded = await this.configManager.loadFromYaml(configPath);
      if (loaded) {
        this.config = deepMerge(COMPLETE_DEFAULTS, this.configManager.getAll());
        
        // 更新转换器的配置
        this.md_to_html = new MarkdownToHtml(this.config);
        this.html_to_word = new HtmlToWordConverter(this.config);
        
        return true;
      }
      return false;
    } catch (error) {
      // console.error(`加载配置文件失败: ${error.message}`);
      return false;
    }
  }

  /**
   * @method cleanup
   * @description 清理临时资源
   */
  cleanup() {
    // 清理可能的临时文件或资源
    // console.log('清理资源...');
  }
}

module.exports = Converter;
